src/gConsent/query/query.ts

```diff
@@ -0,0 +1,262 @@
+ */
+export const DURATION = {
+  ONE_DAY: "P1D",
+  ONE_WEEK: "P7D",
+  ONE_MONTH: "P1M",
+  THREE_MONTHS: "P3M",
+} as const;
+
+export type CredentialFilter = {
+  /**
+   * The Access Credential type (e.g. Access Request, Access Grant...)
+   */
+  type?: CredentialType;
+  /**
+   * The Access Credential status (e.g. Active, Revoked...)
+   */
+  status?: CredentialStatus;
+  /**
+   * WebID of the Agent who issued the Access Credential.
+   */
+  fromAgent?: UrlString;
+  /**
+   * WebID of the Agent who is the Access Credential recipient.
+   */
+  toAgent?: UrlString;
+  /**
+   * URL of the resource featured in the Access Credential.
+   */
+  resource?: UrlString;
+  /**
+   * URL of the Access Credential purpose.
+   */
+  purpose?: UrlString;
+  /**
+   * Interval (expressed using ISO 8601) during which the Credential was issued.
+   */
+  issuedWithin?: "P1D" | "P7D" | "P1M" | "P3M";
+  /**
+   * Interval (expressed using ISO 8601) during which the Credential was revoked.
+   */
+  revokedWithin?: "P1D" | "P7D" | "P1M" | "P3M";
+  /**
+   * Number of items per page.
+   */
+  pageSize?: number;
+  /**
+   * Target page (for result pagination).
+   */
+  page?: string;
+};
+
+const FILTER_ELEMENTS: Array<keyof CredentialFilter> = [
+  "fromAgent",
+  "issuedWithin",
+  "page",
+  "pageSize",
+  "purpose",
+  "resource",
+  "revokedWithin",
+  "status",
+  "toAgent",
+  "type",
+] as const;
+
+// The following ensures that, when we add a value to the FILTER_ELEMENTS array, TS complains
+// if there are missing cases when handling the value.
+type SupportedFilterElements =
+  typeof FILTER_ELEMENTS extends Array<infer E> ? E : never;
+
+// The type assertion is okay in the context of the type guard.
+const isSupportedFilterElement = (x: unknown): x is SupportedFilterElements =>
+  FILTER_ELEMENTS.includes(x as keyof CredentialFilter);
+
+const PAGING_RELS = ["first", "last", "prev", "next"] as const;
+
+export type CredentialResult = {
+  /**
+   * Page of Access Credentials matching the query.
+   */
+  items: DatasetWithId[];
+  /**
+   * First page of query results.
+   */
+  first?: CredentialFilter;
+  /**
+   * Previous page of query results.
+   */
+  prev?: CredentialFilter;
+  /**
+   * Next page of query results.
+   */
+  next?: CredentialFilter;
+  /**
+   * Last page of query results.
+   */
+  last?: CredentialFilter;
+};
+
+function toCredentialFilter(url: URL): CredentialFilter {
+  const filter: CredentialFilter = {};
+  url.searchParams.forEach((value, key) => {
+    if (isSupportedFilterElement(key)) {
+      Object.assign(filter, { [`${key}`]: value });
+    }
+  });
+  return filter;
+}
+
+function hasItems(x: unknown): x is { items: unknown[] } {
+  const candidate = x as { items: unknown };
+  return candidate.items !== undefined && Array.isArray(candidate.items);
+}
+
+function isObjectWithId(x: unknown): x is { id: string } {
+  const candidate = x as { id: string };
+  return typeof candidate.id === "string";
```

Comment:
Does it need to test that the `id` is a non-empty string?

Reply:
Again - this is a type guard, not a validator of the data so I think you can ignore my comment

Reply:
Well it depends on the constraints we put on the target type really, in this case I think that's a good point. I was using it as you describe (for validating data structure only), but here it might be more correct to check the validity of the `id` (specifically, that it is a URL)

Reply:
This function was incidentally removed as part of 87c5e87e8d16028db4f11d0dc54ae488704ad08b